##### Copyright 2019 The TensorFlow Authors.

In [1]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

In [2]:
#@title MIT License
#
# Copyright (c) 2017 François Chollet
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

# 케라스와 텐서플로 허브를 사용한 영화 리뷰 텍스트 분류하기

<table class="tfo-notebook-buttons" align="left">
  <td>     <a target="_blank" href="https://www.tensorflow.org/tutorials/keras/text_classification_with_hub"><img src="https://www.tensorflow.org/images/tf_logo_32px.png">TensorFlow.org에서 보기</a>   </td>
  <td><a target="_blank" href="https://colab.research.google.com/github/tensorflow/docs-l10n/blob/master/site/ko/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png">Google Colab에서 실행</a></td>
  <td><a target="_blank" href="https://github.com/tensorflow/docs-l10n/blob/master/site/ko/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png">GitHub에서 소스 보기</a></td>
  <td><a href="https://storage.googleapis.com/tensorflow_docs/docs-l10n/site/ko/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/download_logo_32px.png">노트북 다운로드</a></td>
  <td><a href="https://tfhub.dev/s?module-type=text-embedding"><img src="https://www.tensorflow.org/images/hub_logo_32px.png">TF Hub 모델 보기</a></td>
</table>

이 노트북은 리뷰의 텍스트를 사용하여 영화 리뷰를 *긍정적* 또는 *부정적*으로 분류합니다. *이진(* 또는 2-클래스 분류인 이 예는 광범위하게 적용할 수 있는 중요한 머신러닝 응용 사례입니다.

이 튜토리얼은 [TensorFlow Hub](https://tfhub.dev) 및 Keras를 사용한 전이 학습의 기본적인 응용을 보여줍니다.

여기서 사용하는 [IMDB 데이터세트](https://www.tensorflow.org/api_docs/python/tf/keras/datasets/imdb)에는 [인터넷 영화 데이터베이스](https://www.imdb.com/)에서 가져온 50,000개의 영화 리뷰 텍스트가 포함되어 있습니다. 훈련용 리뷰 25,000개와 테스트용 리뷰 25,000개로 나뉩니다. 훈련 및 테스트 세트는 *균형을 이룹니다*. 즉, 동일한 수의 긍정적인 리뷰와 부정적인 리뷰가 포함되어 있습니다.

이 노트북은 높은 수준의 API인 [`tf.keras`](https://www.tensorflow.org/guide/keras)를 사용하여 TensorFlow에서 모델을 빌드 및 훈련하고, 단일 코드 줄로 [TFHub](https://tfhub.dev)로부터 훈련된 모델을 로드하기 위한 라이브러리인 [`tensorflow_hub`](https://www.tensorflow.org/hub)를 사용합니다. `tf.keras`를 사용한 고급 텍스트 분류 튜토리얼에 대해서는 [MLCC 텍스트 분류 가이드](https://developers.google.com/machine-learning/guides/text-classification/)를 참조하세요.

In [3]:
!pip install tensorflow-hub
!pip install tensorflow-datasets

In [4]:
import os
import numpy as np

import tensorflow as tf
import tensorflow_hub as hub
import tensorflow_datasets as tfds

print("Version: ", tf.__version__)
print("Eager mode: ", tf.executing_eagerly())
print("Hub version: ", hub.__version__)
print("GPU is", "available" if tf.config.list_physical_devices("GPU") else "NOT AVAILABLE")

2022-12-14 20:07:44.068782: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory
2022-12-14 20:07:44.068900: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory
2022-12-14 20:07:44.068911: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.


Version:  2.11.0
Eager mode:  True
Hub version:  0.12.0
GPU is available


## IMDB 데이터셋 다운로드

IMDB 데이터셋은 [imdb reviews](https://www.tensorflow.org/datasets/catalog/imdb_reviews) 또는 [텐서플로 데이터셋](https://www.tensorflow.org/datasets)(TensorFlow datasets)에 포함되어 있습니다. 다음 코드는 IMDB 데이터셋을 컴퓨터(또는 코랩 런타임)에 다운로드합니다:

In [5]:
# Split the training set into 60% and 40% to end up with 15,000 examples
# for training, 10,000 examples for validation and 25,000 examples for testing.
train_data, validation_data, test_data = tfds.load(
    name="imdb_reviews", 
    split=('train[:60%]', 'train[60%:]', 'test'),
    as_supervised=True)

## 데이터 탐색

잠시 데이터 형태를 알아 보죠. 이 데이터셋의 샘플은 전처리된 정수 배열입니다. 이 정수는 영화 리뷰에 나오는 단어를 나타냅니다. 레이블(label)은 정수 0 또는 1입니다. 0은 부정적인 리뷰이고 1은 긍정적인 리뷰입니다.

처음 10개의 샘플을 출력해 보겠습니다.

In [6]:
train_examples_batch, train_labels_batch = next(iter(train_data.batch(10)))
train_examples_batch

<tf.Tensor: shape=(10,), dtype=string, numpy=
array([b"This was an absolutely terrible movie. Don't be lured in by Christopher Walken or Michael Ironside. Both are great actors, but this must simply be their worst role in history. Even their great acting could not redeem this movie's ridiculous storyline. This movie is an early nineties US propaganda piece. The most pathetic scenes were those when the Columbian rebels were making their cases for revolutions. Maria Conchita Alonso appeared phony, and her pseudo-love affair with Walken was nothing but a pathetic emotional plug in a movie that was devoid of any real meaning. I am disappointed that there are movies like this, ruining actor's like Christopher Walken's good name. I could barely sit through it.",
       b'I have been known to fall asleep during films, but this is usually due to a combination of things including, really tired, being warm and comfortable on the sette and having just eaten a lot. However on this occasion I fell 

처음 10개의 레이블도 출력해 보겠습니다.

In [7]:
train_labels_batch

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([0, 0, 0, 1, 1, 1, 0, 0, 0, 0])>

## 모델 구성

신경망은 층을 쌓아서 만듭니다. 여기에는 세 개의 중요한 구조적 결정이 필요합니다:

- 어떻게 텍스트를 표현할 것인가?
- 모델에서 얼마나 많은 층을 사용할 것인가?
- 각 층에서 얼마나 많은 *은닉 유닛*(hidden unit)을 사용할 것인가?

이 예제의 입력 데이터는 문장으로 구성됩니다. 예측할 레이블은 0 또는 1입니다.

텍스트를 표현하는 한 가지 방법은 문장을 임베딩 벡터로 변환하는 것입니다. 사전 훈련 된 텍스트 임베딩을 첫 번째 레이어로 사용할 수 있으며, 두 가지 이점이 있습니다.

- 텍스트 전처리에 대해 걱정할 필요가 없습니다.
- 전이 학습에 따른 이점이 있습니다.
- 임베딩은 고정 크기이기 때문에 처리 과정이 단순해집니다.

이 예에서는 [google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2)라고 하는 [TensorFlow Hub](https://tfhub.dev)에서 **사전 훈련된 텍스트 임베딩 모델**을 사용합니다.

이 튜토리얼에서 사용할 수 있는 TFHub의 다른 많은 사전 훈련된 텍스트 임베딩이 있습니다.

- [google/nnlm-en-dim128/2](https://tfhub.dev/google/nnlm-en-dim128/2) - [google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2)와 동일한 데이터에 동일한 NNLM 아키텍처로 훈련하지만 임베딩 차원이 더 큽니다. 보다 큰 차원의 임베딩은 작업을 개선할 수 있지만 모델을 훈련하는 데 더 오래 걸릴 수 있습니다.
- [google/nnlm-en-dim128-with-normalization/2](https://tfhub.dev/google/nnlm-en-dim128-with-normalization/2) - [google/nnlm-en-dim128/2](https://tfhub.dev/google/nnlm-en-dim128/2)와 동일하지만 구두점 제거와 같은 추가적인 텍스트 정규화가 있습니다. 이는 작업의 텍스트에 추가 문자나 구두점이 포함된 경우 도움이 될 수 있습니다.
- [google/universal-sentence-encoder/4](https://tfhub.dev/google/universal-sentence-encoder/4) - DAN(deep averaging network) 인코더로 훈련된 512 차원 임베딩을 생성하는 훨씬 더 큰 모델입니다.

그 밖에도 많이 있습니다! TFHub에서 더 많은 [텍스트 임베딩 모델](https://tfhub.dev/s?module-type=text-embedding)을 찾아보세요.

먼저 문장을 임베딩시키기 위해 텐서플로 허브 모델을 사용하는 케라스 층을 만들어 보죠. 그다음 몇 개의 샘플을 입력하여 테스트해 보겠습니다. 입력 텍스트의 길이에 상관없이 임베딩의 출력 크기는 `(num_examples, embedding_dimension)`가 됩니다.

In [8]:
embedding = "https://tfhub.dev/google/nnlm-en-dim50/2"
hub_layer = hub.KerasLayer(embedding, input_shape=[], 
                           dtype=tf.string, trainable=True)
hub_layer(train_examples_batch[:3])

<tf.Tensor: shape=(3, 50), dtype=float32, numpy=
array([[ 0.5423194 , -0.01190171,  0.06337537,  0.0686297 , -0.16776839,
        -0.10581177,  0.168653  , -0.04998823, -0.31148052,  0.07910344,
         0.15442258,  0.01488661,  0.03930155,  0.19772716, -0.12215477,
        -0.04120982, -0.27041087, -0.21922147,  0.26517656, -0.80739075,
         0.25833526, -0.31004202,  0.2868321 ,  0.19433866, -0.29036498,
         0.0386285 , -0.78444123, -0.04793238,  0.41102988, -0.36388886,
        -0.58034706,  0.30269453,  0.36308962, -0.15227163, -0.4439151 ,
         0.19462997,  0.19528405,  0.05666233,  0.2890704 , -0.28468323,
        -0.00531206,  0.0571938 , -0.3201319 , -0.04418665, -0.08550781,
        -0.55847436, -0.2333639 , -0.20782956, -0.03543065, -0.17533456],
       [ 0.56338924, -0.12339553, -0.10862677,  0.7753425 , -0.07667087,
        -0.15752274,  0.01872334, -0.08169781, -0.3521876 ,  0.46373403,
        -0.08492758,  0.07166861, -0.00670818,  0.12686071, -0.19326551,
 

이제 전체 모델을 만들어 보겠습니다:

In [9]:
model = tf.keras.Sequential()
model.add(hub_layer)
model.add(tf.keras.layers.Dense(16, activation='relu'))
model.add(tf.keras.layers.Dense(1))

model.summary()

Instructions for updating:
Lambda fuctions will be no more assumed to be used in the statement where they are used, or at least in the same block. https://github.com/tensorflow/tensorflow/issues/56089


Instructions for updating:
Lambda fuctions will be no more assumed to be used in the statement where they are used, or at least in the same block. https://github.com/tensorflow/tensorflow/issues/56089


Model: "sequential"


_________________________________________________________________


 Layer (type)                Output Shape              Param #   


 keras_layer (KerasLayer)    (None, 50)                48190600  


 dense (Dense)               (None, 16)                816       


 dense_1 (Dense)             (None, 1)                 17        


Total params: 48,191,433


Trainable params: 48,191,433


Non-trainable params: 0


_________________________________________________________________


순서대로 층을 쌓아 분류기를 만듭니다:

1. 첫 번째 레이어는 TensorFlow Hub 레이어입니다. 이 레이어는 사전 훈련된 저장된 모델을 사용하여 문장을 임베딩 벡터에 매핑합니다. 사용 중인 사전 훈련된 텍스트 임베딩 모델([google/nnlm-en-dim50/2](https://tfhub.dev/google/nnlm-en-dim50/2))은 문장을 토큰으로 분할하고 각 토큰을 임베딩한 다음 임베딩을 결합합니다. 결과적인 차원은 `(num_examples, embedding_dimension)`입니다. 이 NNLM 모델의 경우에는 `embedding_dimension`은 50입니다.
2. 이 고정 크기의 출력 벡터는 16개의 은닉 유닛(hidden unit)을 가진 완전 연결 층(`Dense`)으로 주입됩니다.
3. 마지막 층은 하나의 출력 노드를 가진 완전 연결 층입니다. `sigmoid` 활성화 함수를 사용하므로 확률 또는 신뢰도 수준을 표현하는 0~1 사이의 실수가 출력됩니다.

이제 모델을 컴파일합니다.

### 손실 함수와 옵티마이저

모델에는 훈련을 위한 손실 함수와 옵티마이저가 필요합니다. 이진 분류 문제이고 모델이 로짓(선형 활성화가 있는 단일 단위 레이어)을 출력하므로 `binary_crossentropy` 손실 함수를 사용합니다.

다른 손실 함수를 선택할 수 없는 것은 아닙니다. 예를 들어 `mean_squared_error`를 선택할 수 있습니다. 하지만 일반적으로 `binary_crossentropy`가 확률을 다루는데 적합합니다. 이 함수는 확률 분포 간의 거리를 측정합니다. 여기에서는 정답인 타깃 분포와 예측 분포 사이의 거리입니다.

나중에 회귀 문제(예: 주택 가격 예측)를 살펴볼 때 평균 제곱 오차라고 하는 또 다른 손실 함수를 사용하는 방법을 살펴볼 것입니다.

이제 모델이 사용할 옵티마이저와 손실 함수를 설정해 보죠:

In [10]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.BinaryCrossentropy(from_logits=True),
              metrics=['accuracy'])

## 모델 훈련

512개 샘플의 미니 배치에서 10개 epoch 동안 모델을 훈련합니다. 이 동작은 `x_train` 및 `y_train` 텐서의 모든 샘플에 대한 10회 반복에 해당합니다. 훈련하는 동안 검증 세트의 10,000개 샘플에서 모델의 손실과 정확도를 모니터링합니다.

In [11]:
history = model.fit(train_data.shuffle(10000).batch(512),
                    epochs=10,
                    validation_data=validation_data.batch(512),
                    verbose=1)

Epoch 1/10


 1/30 [>.............................] - ETA: 1:59 - loss: 0.6869 - accuracy: 0.5098

 2/30 [=>............................] - ETA: 5s - loss: 0.6839 - accuracy: 0.4941  

 3/30 [==>...........................] - ETA: 5s - loss: 0.6822 - accuracy: 0.4915

 4/30 [===>..........................] - ETA: 5s - loss: 0.6802 - accuracy: 0.5010

 5/30 [====>.........................] - ETA: 4s - loss: 0.6798 - accuracy: 0.5020

 6/30 [=====>........................] - ETA: 4s - loss: 0.6777 - accuracy: 0.5052

 7/30 [======>.......................] - ETA: 4s - loss: 0.6769 - accuracy: 0.5056

 8/30 [=======>......................] - ETA: 4s - loss: 0.6754 - accuracy: 0.5029

 9/30 [========>.....................] - ETA: 4s - loss: 0.6743 - accuracy: 0.5052

10/30 [=========>....................] - ETA: 3s - loss: 0.6730 - accuracy: 0.5045

11/30 [==========>...................] - ETA: 3s - loss: 0.6721 - accuracy: 0.5007

12/30 [===========>..................] - ETA: 3s - loss: 0.6706 - accuracy: 0.5033

13/30 [============>.................] - ETA: 3s - loss: 0.6697 - accuracy: 0.5077

14/30 [=============>................] - ETA: 3s - loss: 0.6690 - accuracy: 0.5070

15/30 [==============>...............] - ETA: 2s - loss: 0.6677 - accuracy: 0.5072

16/30 [===============>..............] - ETA: 2s - loss: 0.6661 - accuracy: 0.5117

17/30 [================>.............] - ETA: 2s - loss: 0.6648 - accuracy: 0.5117

18/30 [=================>............] - ETA: 2s - loss: 0.6631 - accuracy: 0.5145

19/30 [==================>...........] - ETA: 2s - loss: 0.6627 - accuracy: 0.5150

20/30 [===================>..........] - ETA: 1s - loss: 0.6617 - accuracy: 0.5160

21/30 [====================>.........] - ETA: 1s - loss: 0.6605 - accuracy: 0.5166

22/30 [=====================>........] - ETA: 1s - loss: 0.6591 - accuracy: 0.5176

23/30 [======================>.......] - ETA: 1s - loss: 0.6580 - accuracy: 0.5191

24/30 [=======================>......] - ETA: 1s - loss: 0.6569 - accuracy: 0.5210

25/30 [========================>.....] - ETA: 0s - loss: 0.6560 - accuracy: 0.5222

26/30 [=========================>....] - ETA: 0s - loss: 0.6545 - accuracy: 0.5241

27/30 [==========================>...] - ETA: 0s - loss: 0.6533 - accuracy: 0.5263

28/30 [===========================>..] - ETA: 0s - loss: 0.6516 - accuracy: 0.5290

29/30 [============================>.] - ETA: 0s - loss: 0.6503 - accuracy: 0.5315

30/30 [==============================] - ETA: 0s - loss: 0.6499 - accuracy: 0.5321

30/30 [==============================] - 10s 219ms/step - loss: 0.6499 - accuracy: 0.5321 - val_loss: 0.6091 - val_accuracy: 0.5972


Epoch 2/10


 1/30 [>.............................] - ETA: 18s - loss: 0.5909 - accuracy: 0.6055

 2/30 [=>............................] - ETA: 5s - loss: 0.5952 - accuracy: 0.6084 

 3/30 [==>...........................] - ETA: 4s - loss: 0.5957 - accuracy: 0.6068

 4/30 [===>..........................] - ETA: 4s - loss: 0.5946 - accuracy: 0.6084

 5/30 [====>.........................] - ETA: 4s - loss: 0.5940 - accuracy: 0.6102

 6/30 [=====>........................] - ETA: 4s - loss: 0.5915 - accuracy: 0.6217

 7/30 [======>.......................] - ETA: 4s - loss: 0.5889 - accuracy: 0.6292

 8/30 [=======>......................] - ETA: 4s - loss: 0.5858 - accuracy: 0.6353

10/30 [=========>....................] - ETA: 3s - loss: 0.5836 - accuracy: 0.6430

11/30 [==========>...................] - ETA: 3s - loss: 0.5810 - accuracy: 0.6426

12/30 [===========>..................] - ETA: 3s - loss: 0.5792 - accuracy: 0.6436

13/30 [============>.................] - ETA: 2s - loss: 0.5788 - accuracy: 0.6430

14/30 [=============>................] - ETA: 2s - loss: 0.5772 - accuracy: 0.6452

15/30 [==============>...............] - ETA: 2s - loss: 0.5760 - accuracy: 0.6471

16/30 [===============>..............] - ETA: 2s - loss: 0.5748 - accuracy: 0.6482

17/30 [================>.............] - ETA: 2s - loss: 0.5724 - accuracy: 0.6502

18/30 [=================>............] - ETA: 2s - loss: 0.5698 - accuracy: 0.6519

20/30 [===================>..........] - ETA: 1s - loss: 0.5676 - accuracy: 0.6540

21/30 [====================>.........] - ETA: 1s - loss: 0.5657 - accuracy: 0.6577

22/30 [=====================>........] - ETA: 1s - loss: 0.5647 - accuracy: 0.6576

23/30 [======================>.......] - ETA: 1s - loss: 0.5625 - accuracy: 0.6597

24/30 [=======================>......] - ETA: 1s - loss: 0.5603 - accuracy: 0.6630

25/30 [========================>.....] - ETA: 0s - loss: 0.5592 - accuracy: 0.6645

26/30 [=========================>....] - ETA: 0s - loss: 0.5577 - accuracy: 0.6675

27/30 [==========================>...] - ETA: 0s - loss: 0.5559 - accuracy: 0.6711

28/30 [===========================>..] - ETA: 0s - loss: 0.5546 - accuracy: 0.6735

29/30 [============================>.] - ETA: 0s - loss: 0.5529 - accuracy: 0.6761

30/30 [==============================] - ETA: 0s - loss: 0.5530 - accuracy: 0.6763

30/30 [==============================] - 6s 197ms/step - loss: 0.5530 - accuracy: 0.6763 - val_loss: 0.5170 - val_accuracy: 0.7464


Epoch 3/10


 1/30 [>.............................] - ETA: 18s - loss: 0.4565 - accuracy: 0.8105

 2/30 [=>............................] - ETA: 5s - loss: 0.4675 - accuracy: 0.8047 

 4/30 [===>..........................] - ETA: 3s - loss: 0.4739 - accuracy: 0.7861

 5/30 [====>.........................] - ETA: 3s - loss: 0.4749 - accuracy: 0.7789

 6/30 [=====>........................] - ETA: 3s - loss: 0.4692 - accuracy: 0.7796

 7/30 [======>.......................] - ETA: 3s - loss: 0.4656 - accuracy: 0.7812

 8/30 [=======>......................] - ETA: 3s - loss: 0.4666 - accuracy: 0.7761

 9/30 [========>.....................] - ETA: 3s - loss: 0.4648 - accuracy: 0.7760

10/30 [=========>....................] - ETA: 3s - loss: 0.4623 - accuracy: 0.7770

11/30 [==========>...................] - ETA: 3s - loss: 0.4599 - accuracy: 0.7802

12/30 [===========>..................] - ETA: 3s - loss: 0.4594 - accuracy: 0.7826

13/30 [============>.................] - ETA: 2s - loss: 0.4585 - accuracy: 0.7852

14/30 [=============>................] - ETA: 2s - loss: 0.4557 - accuracy: 0.7891

16/30 [===============>..............] - ETA: 2s - loss: 0.4530 - accuracy: 0.7917

17/30 [================>.............] - ETA: 2s - loss: 0.4512 - accuracy: 0.7918

18/30 [=================>............] - ETA: 2s - loss: 0.4483 - accuracy: 0.7954

19/30 [==================>...........] - ETA: 1s - loss: 0.4475 - accuracy: 0.7963

20/30 [===================>..........] - ETA: 1s - loss: 0.4455 - accuracy: 0.7973

21/30 [====================>.........] - ETA: 1s - loss: 0.4438 - accuracy: 0.7975

22/30 [=====================>........] - ETA: 1s - loss: 0.4416 - accuracy: 0.7984

23/30 [======================>.......] - ETA: 1s - loss: 0.4399 - accuracy: 0.7993

24/30 [=======================>......] - ETA: 1s - loss: 0.4376 - accuracy: 0.8012

25/30 [========================>.....] - ETA: 0s - loss: 0.4364 - accuracy: 0.8019

26/30 [=========================>....] - ETA: 0s - loss: 0.4351 - accuracy: 0.8027

27/30 [==========================>...] - ETA: 0s - loss: 0.4341 - accuracy: 0.8038

28/30 [===========================>..] - ETA: 0s - loss: 0.4335 - accuracy: 0.8030

29/30 [============================>.] - ETA: 0s - loss: 0.4313 - accuracy: 0.8048

30/30 [==============================] - ETA: 0s - loss: 0.4304 - accuracy: 0.8055

30/30 [==============================] - 7s 203ms/step - loss: 0.4304 - accuracy: 0.8055 - val_loss: 0.4168 - val_accuracy: 0.8126


Epoch 4/10


 1/30 [>.............................] - ETA: 18s - loss: 0.3653 - accuracy: 0.8535

 2/30 [=>............................] - ETA: 5s - loss: 0.3618 - accuracy: 0.8564 

 3/30 [==>...........................] - ETA: 5s - loss: 0.3586 - accuracy: 0.8594

 4/30 [===>..........................] - ETA: 5s - loss: 0.3603 - accuracy: 0.8491

 5/30 [====>.........................] - ETA: 5s - loss: 0.3512 - accuracy: 0.8535

 6/30 [=====>........................] - ETA: 4s - loss: 0.3493 - accuracy: 0.8555

 7/30 [======>.......................] - ETA: 4s - loss: 0.3477 - accuracy: 0.8588

 8/30 [=======>......................] - ETA: 4s - loss: 0.3480 - accuracy: 0.8552

 9/30 [========>.....................] - ETA: 4s - loss: 0.3422 - accuracy: 0.8585

10/30 [=========>....................] - ETA: 3s - loss: 0.3397 - accuracy: 0.8611

11/30 [==========>...................] - ETA: 3s - loss: 0.3386 - accuracy: 0.8624

12/30 [===========>..................] - ETA: 3s - loss: 0.3352 - accuracy: 0.8644

13/30 [============>.................] - ETA: 3s - loss: 0.3332 - accuracy: 0.8652

14/30 [=============>................] - ETA: 3s - loss: 0.3338 - accuracy: 0.8652

15/30 [==============>...............] - ETA: 2s - loss: 0.3316 - accuracy: 0.8660

16/30 [===============>..............] - ETA: 2s - loss: 0.3297 - accuracy: 0.8665

17/30 [================>.............] - ETA: 2s - loss: 0.3282 - accuracy: 0.8668

18/30 [=================>............] - ETA: 2s - loss: 0.3259 - accuracy: 0.8671

19/30 [==================>...........] - ETA: 2s - loss: 0.3247 - accuracy: 0.8672

20/30 [===================>..........] - ETA: 1s - loss: 0.3245 - accuracy: 0.8681

21/30 [====================>.........] - ETA: 1s - loss: 0.3236 - accuracy: 0.8688

22/30 [=====================>........] - ETA: 1s - loss: 0.3233 - accuracy: 0.8695

23/30 [======================>.......] - ETA: 1s - loss: 0.3211 - accuracy: 0.8710

25/30 [========================>.....] - ETA: 0s - loss: 0.3181 - accuracy: 0.8725

26/30 [=========================>....] - ETA: 0s - loss: 0.3181 - accuracy: 0.8721

27/30 [==========================>...] - ETA: 0s - loss: 0.3167 - accuracy: 0.8728

28/30 [===========================>..] - ETA: 0s - loss: 0.3155 - accuracy: 0.8727

29/30 [============================>.] - ETA: 0s - loss: 0.3157 - accuracy: 0.8720

30/30 [==============================] - ETA: 0s - loss: 0.3149 - accuracy: 0.8724

30/30 [==============================] - 7s 207ms/step - loss: 0.3149 - accuracy: 0.8724 - val_loss: 0.3539 - val_accuracy: 0.8494


Epoch 5/10


 1/30 [>.............................] - ETA: 20s - loss: 0.2565 - accuracy: 0.9082

 2/30 [=>............................] - ETA: 5s - loss: 0.2548 - accuracy: 0.9121 

 3/30 [==>...........................] - ETA: 5s - loss: 0.2612 - accuracy: 0.9115

 5/30 [====>.........................] - ETA: 3s - loss: 0.2587 - accuracy: 0.9125

 6/30 [=====>........................] - ETA: 3s - loss: 0.2569 - accuracy: 0.9131

 7/30 [======>.......................] - ETA: 3s - loss: 0.2535 - accuracy: 0.9146

 8/30 [=======>......................] - ETA: 3s - loss: 0.2530 - accuracy: 0.9131

 9/30 [========>.....................] - ETA: 3s - loss: 0.2521 - accuracy: 0.9132

10/30 [=========>....................] - ETA: 3s - loss: 0.2488 - accuracy: 0.9150

11/30 [==========>...................] - ETA: 3s - loss: 0.2477 - accuracy: 0.9141

12/30 [===========>..................] - ETA: 3s - loss: 0.2459 - accuracy: 0.9144

13/30 [============>.................] - ETA: 2s - loss: 0.2462 - accuracy: 0.9132

15/30 [==============>...............] - ETA: 2s - loss: 0.2465 - accuracy: 0.9135

16/30 [===============>..............] - ETA: 2s - loss: 0.2451 - accuracy: 0.9137

17/30 [================>.............] - ETA: 2s - loss: 0.2435 - accuracy: 0.9136

18/30 [=================>............] - ETA: 2s - loss: 0.2413 - accuracy: 0.9150

21/30 [====================>.........] - ETA: 1s - loss: 0.2386 - accuracy: 0.9166

22/30 [=====================>........] - ETA: 1s - loss: 0.2372 - accuracy: 0.9169

23/30 [======================>.......] - ETA: 1s - loss: 0.2369 - accuracy: 0.9164

24/30 [=======================>......] - ETA: 0s - loss: 0.2367 - accuracy: 0.9163

26/30 [=========================>....] - ETA: 0s - loss: 0.2340 - accuracy: 0.9163

27/30 [==========================>...] - ETA: 0s - loss: 0.2333 - accuracy: 0.9165

28/30 [===========================>..] - ETA: 0s - loss: 0.2320 - accuracy: 0.9170

30/30 [==============================] - ETA: 0s - loss: 0.2311 - accuracy: 0.9173

30/30 [==============================] - 6s 180ms/step - loss: 0.2311 - accuracy: 0.9173 - val_loss: 0.3240 - val_accuracy: 0.8619


Epoch 6/10


 1/30 [>.............................] - ETA: 20s - loss: 0.2061 - accuracy: 0.9258

 2/30 [=>............................] - ETA: 5s - loss: 0.2011 - accuracy: 0.9346 

 3/30 [==>...........................] - ETA: 5s - loss: 0.1857 - accuracy: 0.9460

 4/30 [===>..........................] - ETA: 4s - loss: 0.1857 - accuracy: 0.9453

 5/30 [====>.........................] - ETA: 4s - loss: 0.1822 - accuracy: 0.9461

 6/30 [=====>........................] - ETA: 4s - loss: 0.1832 - accuracy: 0.9437

 8/30 [=======>......................] - ETA: 3s - loss: 0.1801 - accuracy: 0.9407

 9/30 [========>.....................] - ETA: 3s - loss: 0.1796 - accuracy: 0.9403

11/30 [==========>...................] - ETA: 2s - loss: 0.1773 - accuracy: 0.9421

13/30 [============>.................] - ETA: 2s - loss: 0.1763 - accuracy: 0.9431

14/30 [=============>................] - ETA: 2s - loss: 0.1754 - accuracy: 0.9432

15/30 [==============>...............] - ETA: 2s - loss: 0.1751 - accuracy: 0.9431

17/30 [================>.............] - ETA: 1s - loss: 0.1741 - accuracy: 0.9429

18/30 [=================>............] - ETA: 1s - loss: 0.1741 - accuracy: 0.9427

19/30 [==================>...........] - ETA: 1s - loss: 0.1753 - accuracy: 0.9422

20/30 [===================>..........] - ETA: 1s - loss: 0.1749 - accuracy: 0.9427

21/30 [====================>.........] - ETA: 1s - loss: 0.1737 - accuracy: 0.9435

22/30 [=====================>........] - ETA: 1s - loss: 0.1748 - accuracy: 0.9427

23/30 [======================>.......] - ETA: 1s - loss: 0.1749 - accuracy: 0.9427

25/30 [========================>.....] - ETA: 0s - loss: 0.1742 - accuracy: 0.9428

26/30 [=========================>....] - ETA: 0s - loss: 0.1745 - accuracy: 0.9430

27/30 [==========================>...] - ETA: 0s - loss: 0.1739 - accuracy: 0.9430

28/30 [===========================>..] - ETA: 0s - loss: 0.1734 - accuracy: 0.9433

29/30 [============================>.] - ETA: 0s - loss: 0.1725 - accuracy: 0.9437

30/30 [==============================] - ETA: 0s - loss: 0.1718 - accuracy: 0.9442

30/30 [==============================] - 6s 181ms/step - loss: 0.1718 - accuracy: 0.9442 - val_loss: 0.3088 - val_accuracy: 0.8657


Epoch 7/10


 1/30 [>.............................] - ETA: 25s - loss: 0.1169 - accuracy: 0.9668

 2/30 [=>............................] - ETA: 5s - loss: 0.1189 - accuracy: 0.9678 

 3/30 [==>...........................] - ETA: 5s - loss: 0.1294 - accuracy: 0.9642

 4/30 [===>..........................] - ETA: 5s - loss: 0.1314 - accuracy: 0.9644

 6/30 [=====>........................] - ETA: 3s - loss: 0.1303 - accuracy: 0.9609

 8/30 [=======>......................] - ETA: 3s - loss: 0.1286 - accuracy: 0.9619

 9/30 [========>.....................] - ETA: 3s - loss: 0.1286 - accuracy: 0.9622

10/30 [=========>....................] - ETA: 3s - loss: 0.1278 - accuracy: 0.9635

11/30 [==========>...................] - ETA: 3s - loss: 0.1262 - accuracy: 0.9648

12/30 [===========>..................] - ETA: 2s - loss: 0.1258 - accuracy: 0.9645

13/30 [============>.................] - ETA: 2s - loss: 0.1264 - accuracy: 0.9651

14/30 [=============>................] - ETA: 2s - loss: 0.1256 - accuracy: 0.9651

15/30 [==============>...............] - ETA: 2s - loss: 0.1266 - accuracy: 0.9646

16/30 [===============>..............] - ETA: 2s - loss: 0.1273 - accuracy: 0.9637

18/30 [=================>............] - ETA: 1s - loss: 0.1274 - accuracy: 0.9626

19/30 [==================>...........] - ETA: 1s - loss: 0.1271 - accuracy: 0.9626

20/30 [===================>..........] - ETA: 1s - loss: 0.1277 - accuracy: 0.9623

21/30 [====================>.........] - ETA: 1s - loss: 0.1272 - accuracy: 0.9623

22/30 [=====================>........] - ETA: 1s - loss: 0.1273 - accuracy: 0.9617

23/30 [======================>.......] - ETA: 1s - loss: 0.1272 - accuracy: 0.9618

24/30 [=======================>......] - ETA: 1s - loss: 0.1273 - accuracy: 0.9620

27/30 [==========================>...] - ETA: 0s - loss: 0.1275 - accuracy: 0.9616

28/30 [===========================>..] - ETA: 0s - loss: 0.1272 - accuracy: 0.9615

30/30 [==============================] - ETA: 0s - loss: 0.1275 - accuracy: 0.9613

30/30 [==============================] - 6s 181ms/step - loss: 0.1275 - accuracy: 0.9613 - val_loss: 0.3061 - val_accuracy: 0.8709


Epoch 8/10


 1/30 [>.............................] - ETA: 21s - loss: 0.1056 - accuracy: 0.9688

 2/30 [=>............................] - ETA: 5s - loss: 0.1116 - accuracy: 0.9648 

 3/30 [==>...........................] - ETA: 5s - loss: 0.1044 - accuracy: 0.9727

 4/30 [===>..........................] - ETA: 5s - loss: 0.1043 - accuracy: 0.9731

 5/30 [====>.........................] - ETA: 4s - loss: 0.1037 - accuracy: 0.9723

 6/30 [=====>........................] - ETA: 4s - loss: 0.1033 - accuracy: 0.9727

 7/30 [======>.......................] - ETA: 4s - loss: 0.1032 - accuracy: 0.9724

 9/30 [========>.....................] - ETA: 3s - loss: 0.0985 - accuracy: 0.9737

10/30 [=========>....................] - ETA: 3s - loss: 0.0978 - accuracy: 0.9744

11/30 [==========>...................] - ETA: 3s - loss: 0.0963 - accuracy: 0.9744

12/30 [===========>..................] - ETA: 3s - loss: 0.0957 - accuracy: 0.9746

13/30 [============>.................] - ETA: 3s - loss: 0.0960 - accuracy: 0.9745

14/30 [=============>................] - ETA: 2s - loss: 0.0955 - accuracy: 0.9745

15/30 [==============>...............] - ETA: 2s - loss: 0.0951 - accuracy: 0.9751

16/30 [===============>..............] - ETA: 2s - loss: 0.0933 - accuracy: 0.9757

17/30 [================>.............] - ETA: 2s - loss: 0.0925 - accuracy: 0.9762

18/30 [=================>............] - ETA: 2s - loss: 0.0920 - accuracy: 0.9762

19/30 [==================>...........] - ETA: 2s - loss: 0.0921 - accuracy: 0.9760

20/30 [===================>..........] - ETA: 1s - loss: 0.0935 - accuracy: 0.9760

22/30 [=====================>........] - ETA: 1s - loss: 0.0938 - accuracy: 0.9759

23/30 [======================>.......] - ETA: 1s - loss: 0.0940 - accuracy: 0.9758

24/30 [=======================>......] - ETA: 1s - loss: 0.0931 - accuracy: 0.9764

27/30 [==========================>...] - ETA: 0s - loss: 0.0918 - accuracy: 0.9761

29/30 [============================>.] - ETA: 0s - loss: 0.0911 - accuracy: 0.9764

30/30 [==============================] - ETA: 0s - loss: 0.0914 - accuracy: 0.9763

30/30 [==============================] - 6s 182ms/step - loss: 0.0914 - accuracy: 0.9763 - val_loss: 0.3104 - val_accuracy: 0.8702


Epoch 9/10


 1/30 [>.............................] - ETA: 20s - loss: 0.0717 - accuracy: 0.9883

 2/30 [=>............................] - ETA: 5s - loss: 0.0730 - accuracy: 0.9873 

 3/30 [==>...........................] - ETA: 5s - loss: 0.0742 - accuracy: 0.9811

 4/30 [===>..........................] - ETA: 4s - loss: 0.0671 - accuracy: 0.9849

 5/30 [====>.........................] - ETA: 4s - loss: 0.0669 - accuracy: 0.9855

 6/30 [=====>........................] - ETA: 4s - loss: 0.0668 - accuracy: 0.9860

 7/30 [======>.......................] - ETA: 4s - loss: 0.0670 - accuracy: 0.9866

 8/30 [=======>......................] - ETA: 4s - loss: 0.0671 - accuracy: 0.9868

 9/30 [========>.....................] - ETA: 3s - loss: 0.0667 - accuracy: 0.9865

11/30 [==========>...................] - ETA: 3s - loss: 0.0662 - accuracy: 0.9862

12/30 [===========>..................] - ETA: 3s - loss: 0.0670 - accuracy: 0.9855

13/30 [============>.................] - ETA: 2s - loss: 0.0668 - accuracy: 0.9853

14/30 [=============>................] - ETA: 2s - loss: 0.0672 - accuracy: 0.9854

15/30 [==============>...............] - ETA: 2s - loss: 0.0675 - accuracy: 0.9859

17/30 [================>.............] - ETA: 2s - loss: 0.0669 - accuracy: 0.9861

18/30 [=================>............] - ETA: 1s - loss: 0.0667 - accuracy: 0.9865

19/30 [==================>...........] - ETA: 1s - loss: 0.0675 - accuracy: 0.9860

20/30 [===================>..........] - ETA: 1s - loss: 0.0673 - accuracy: 0.9860

21/30 [====================>.........] - ETA: 1s - loss: 0.0672 - accuracy: 0.9860

22/30 [=====================>........] - ETA: 1s - loss: 0.0674 - accuracy: 0.9854

25/30 [========================>.....] - ETA: 0s - loss: 0.0669 - accuracy: 0.9854

26/30 [=========================>....] - ETA: 0s - loss: 0.0665 - accuracy: 0.9855

27/30 [==========================>...] - ETA: 0s - loss: 0.0661 - accuracy: 0.9855

29/30 [============================>.] - ETA: 0s - loss: 0.0660 - accuracy: 0.9856

30/30 [==============================] - ETA: 0s - loss: 0.0660 - accuracy: 0.9856

30/30 [==============================] - 6s 178ms/step - loss: 0.0660 - accuracy: 0.9856 - val_loss: 0.3211 - val_accuracy: 0.8714


Epoch 10/10


 1/30 [>.............................] - ETA: 14s - loss: 0.0566 - accuracy: 0.9824

 2/30 [=>............................] - ETA: 5s - loss: 0.0519 - accuracy: 0.9883 

 3/30 [==>...........................] - ETA: 5s - loss: 0.0522 - accuracy: 0.9889

 4/30 [===>..........................] - ETA: 4s - loss: 0.0540 - accuracy: 0.9893

 5/30 [====>.........................] - ETA: 4s - loss: 0.0526 - accuracy: 0.9902

 6/30 [=====>........................] - ETA: 4s - loss: 0.0519 - accuracy: 0.9909

 7/30 [======>.......................] - ETA: 4s - loss: 0.0509 - accuracy: 0.9914

 8/30 [=======>......................] - ETA: 4s - loss: 0.0503 - accuracy: 0.9915

 9/30 [========>.....................] - ETA: 3s - loss: 0.0496 - accuracy: 0.9915

10/30 [=========>....................] - ETA: 3s - loss: 0.0493 - accuracy: 0.9918

11/30 [==========>...................] - ETA: 3s - loss: 0.0490 - accuracy: 0.9918

13/30 [============>.................] - ETA: 2s - loss: 0.0481 - accuracy: 0.9916

14/30 [=============>................] - ETA: 2s - loss: 0.0475 - accuracy: 0.9918

16/30 [===============>..............] - ETA: 2s - loss: 0.0489 - accuracy: 0.9916

17/30 [================>.............] - ETA: 2s - loss: 0.0486 - accuracy: 0.9917

18/30 [=================>............] - ETA: 1s - loss: 0.0485 - accuracy: 0.9919

19/30 [==================>...........] - ETA: 1s - loss: 0.0486 - accuracy: 0.9918

20/30 [===================>..........] - ETA: 1s - loss: 0.0483 - accuracy: 0.9917

21/30 [====================>.........] - ETA: 1s - loss: 0.0479 - accuracy: 0.9918

22/30 [=====================>........] - ETA: 1s - loss: 0.0474 - accuracy: 0.9918

24/30 [=======================>......] - ETA: 0s - loss: 0.0482 - accuracy: 0.9917

25/30 [========================>.....] - ETA: 0s - loss: 0.0477 - accuracy: 0.9917

26/30 [=========================>....] - ETA: 0s - loss: 0.0479 - accuracy: 0.9915

27/30 [==========================>...] - ETA: 0s - loss: 0.0474 - accuracy: 0.9917

28/30 [===========================>..] - ETA: 0s - loss: 0.0474 - accuracy: 0.9917

30/30 [==============================] - ETA: 0s - loss: 0.0468 - accuracy: 0.9920

30/30 [==============================] - 6s 183ms/step - loss: 0.0468 - accuracy: 0.9920 - val_loss: 0.3323 - val_accuracy: 0.8686


## 모델 평가

모델의 성능을 확인해 보죠. 두 개의 값이 반환됩니다. 손실(오차를 나타내는 숫자이므로 낮을수록 좋습니다)과 정확도입니다.

In [12]:
results = model.evaluate(test_data.batch(512), verbose=2)

for name, value in zip(model.metrics_names, results):
  print("%s: %.3f" % (name, value))

49/49 - 1s - loss: 0.3630 - accuracy: 0.8524 - 1s/epoch - 24ms/step


loss: 0.363
accuracy: 0.852


이 예제는 매우 단순한 방식을 사용하므로 87% 정도의 정확도를 달성했습니다. 고급 방법을 사용한 모델은 95%에 가까운 정확도를 얻습니다.

## 더 읽을거리

- 문자열 입력으로 작업하는 보다 일반적인 방법과 훈련 중 정확도 및 손실의 진행 상황에 대한 보다 자세한 분석은 [전처리된 텍스트를 사용한 텍스트 분류](./text_classification.ipynb) 튜토리얼을 참조하세요.
- TFHub에서 훈련된 모델을 사용하여 더 많은 [텍스트 관련 튜토리얼](https://www.tensorflow.org/hub/tutorials#text-related-tutorials)을 시도해 보세요.